Objects/rangeobject.c

```diff
@@ -527,35 +527,101 @@ range_count(rangeobject *r, PyObject *ob)
 }
 
 static PyObject *
-range_index(rangeobject *r, PyObject *ob)
+range_index(rangeobject *r, PyObject *args, PyObject *kw)
 {
     int contains;
+    static char *keywords[] = {"value", "start", "stop", NULL};
+    PyObject *ob;               /* ob is the key */
+    PyObject *start = NULL;
+    PyObject *stop = NULL;
+    if(!PyArg_ParseTupleAndKeywords(args, kw, "O|OO:range_index",
+                                    keywords, &ob, &start, &stop)){
+      return NULL;
+    }
 
+    PyObject *return_value;
+    if(start || stop){
+      /* This means we have to use only a portion of the range. We
+         create a relavant slice of the range and do the indexing on
+         that slice */
+        if(start){
+            if(!PyLong_CheckExact(start)){
+                PyErr_Format(PyExc_TypeError,
+                             "start value must be integer, not %.200s",
+                             start->ob_type->tp_name);
+                return NULL;
+            }
+        }
+        if(stop){
+            if(!PyLong_CheckExact(stop)){
+                PyErr_Format(PyExc_TypeError,
+                             "stop value must be integer, not %.200s",
+                             stop->ob_type->tp_name);
+                return NULL;
+            }
+        }
```

Comment:
Make the code look a little bit cleaner.
```suggestion
        if(start && (!PyLong_CheckExact(start))){
            PyErr_Format(PyExc_TypeError,
                         "start value must be integer, not %.200s",
                         start->ob_type->tp_name);
            return NULL;
        }         
        if(stop && (!PyLong_CheckExact(stop))){
            PyErr_Format(PyExc_TypeError,
                         "stop value must be integer, not %.200s",
                         stop->ob_type->tp_name);
            return NULL;
        } 
```